notebooks/Batchelor.ipynb

```diff
@@ -0,0 +1,348 @@
+    "    geom = pyvista.Arrow()\n",
+    "    glyphs = grid.glyph(orient=\"u\", geom=geom, factor=0.09)\n",
+    "\n",
+    "    plotter = pyvista.Plotter(window_size=[800,800])\n",
+    "    #plotter.add_mesh(grid, show_edges=True, show_scalar_bar=False, cmap='coolwarm')\n",
+    "    plotter.add_mesh(glyphs, cmap='coolwarm', show_scalar_bar=True)\n",
+    "    plotter.add_text(\"Batchelor Cornerflow Approximated Solution\", position=\"upper_edge\", font_size=14, color=\"black\")\n",
+    "    plotter.view_xy()\n",
+    "    plotter.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7b488e33-8759-4d1c-98dc-13c7e9ccfb39",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def convergence():\n",
+    "    fig = pl.figure() # open plotting figures\n",
+    "    fig2 = pl.figure()\n",
+    "    \n",
+    "    ax = fig.gca()\n",
+    "    ax2 = fig2.gca()\n",
+    "    \n",
+    "    ps = [1, 2] # polynomial orders to try\n",
+    "\n",
+    "    nelements = [10, 20, 40, 80] # resolutions to try\n",
+    "\n",
+    "    test_passes = True # expected order of convergence flag\n",
+    "\n",
+    "    for p in ps: # loop over polynomial orders\n",
+    "        errors_l2_a = [] # accumulate l2 errors\n",
+    "        errors_max = [] # accumulate max errors\n",
+    "        for ne in nelements: # loop over resolutions\n",
+    "            vh = batchelor(ne, p) # solve 2D Poisson\n",
+    "            errorl2, errormax = error(vh) # evalute approximate solution error\n",
+    "            print('ne = ', ne, ', l2error = ', errorl2) # print and save l2 error\n",
+    "            print('ne = ', ne, ', max error = ', errormax) # print and save max error\n",
+    "            errors_l2_a.append(errorl2)\n",
+    "            errors_max.append(errormax)\n",
+    "        hs = 1./np.array(nelements)/p #work out order of convergence\n",
+    "        fit = np.polyfit(np.log(hs), np.log(errors_l2_a),1) # fit line to convergence data for l2 error\n",
+    "        fit2 = np.polyfit(np.log(hs), np.log(errors_max),1) # fit line to convergence data for max error\n",
+    "        print(\"***********  order of accuracy p={}, order={}\".format(p,fit[0]))\n",
+    "\n",
+    "        # log-log plot of the L2 error \n",
+    "        ax.loglog(hs,errors_l2_a,'o-',label='p={}, order={:.2f}'.format(p,fit[0]))\n",
+    "        # log-log plot of the max error\n",
+    "        ax2.loglog(hs,errors_max,'o-',label='p={}, order={:.2f}'.format(p,fit2[0]))\n",
+    "    \n",
+    "        # Test if the order of convergence is as expected\n",
+    "        test_passes = test_passes and abs(fit[0]-1) < 0.1\n",
+    "\n",
+    "    # Tidy up the plot\n",
+    "    %matplotlib inline\n",
+    "    ax.set_xlabel('h')\n",
+    "    ax.set_ylabel('||e||_2')\n",
+    "    ax.grid()\n",
+    "    ax.set_title('Convergence: L2 Error')\n",
+    "    ax.legend()\n",
+    "    fig.savefig('batchelor_cornerflow.pdf')\n",
```

Comment:
Let's also talk about adding an output folder to this so that this file isn't just put into the notebooks directory.